core/coreapi/unixfs.go

```diff
@@ -367,3 +378,33 @@ type syncDagService struct {
 func (s *syncDagService) Sync() error {
 	return s.syncFn()
 }
+
+type providingDagService struct {
+	ipld.DAGService
+	provider provider.System
+}
+
+func (pds *providingDagService) Add(ctx context.Context, n ipld.Node) error {
+	if err := pds.DAGService.Add(ctx, n); err != nil {
+		return err
+	}
+	if err := pds.provider.Provide(ctx, n.Cid(), true); err != nil {
+		log.Error(err)
+	}
+	return nil
+}
+
+func (pds *providingDagService) AddMany(ctx context.Context, nds []ipld.Node) error {
+	if err := pds.DAGService.AddMany(ctx, nds); err != nil {
+		return err
+	}
+	for _, n := range nds {
+		if err := pds.provider.Provide(ctx, n.Cid(), true); err != nil {
+			log.Error(err)
+			break
+		}
```

Comment:
The fact that this a reprovider means it will internally use ProvideMany on the real provider when possible. ProvideMany() is not available to us here directly. Plus with upcoming changes it has no point to bring it up here.